README.md
# CMS-Drupal-Modules-MembershipEntity
Perl interface and utilities for the Drupal MembershipEntity modules

You may like to use this software if you manage a Drupal website in which you use
Membership Entity to manage your site's memberships.


Changes
{{$NEXT}}

Don't use open.pm with any version of Perl 5.14
on MSWin32

0.95      2015-08-02 22:39:04-07:00 America/Los_Angeles

Don't use open.pm with Strawberry Perl 5.14.3
(bug: see http://www.perlmonks.org/?node_id=1136800)

Add prereq DBD::SQLite

Add minimum Perl version to Makefile.PM with
Perl::MinimumVersionFast dzil plugin

0.94      2015-07-28 20:06:47-07:00 America/Los_Angeles

Bump version to force retest on CPAN; failure was
reported on t/40_fetch_memberships.t but it looks
bogus...

0.93      2015-07-26 15:57:50-07:00 America/Los_Angeles

Implement Test::Exception in the test suite

Return the object directly when fetch_memberships() is
called with a single mid

Add historical count of renewal memberships

0.92      2015-07-25 15:31:43-07:00 America/Los_Angeles

Methods to provide counts from past dates

Utility methods in Stats.pm

Add tutorial Cookbook.pm

0.91      2015-07-14 09:26:24-07:00 America/Los_Angeles

Allow use of local Drupal DB in some tests

Take an array as argument to fetch_memberships

0.90      2015-07-14 05:40:04-07:00 America/Los_Angeles

Introduce basic Stats package and tests

0.095     2015-07-12 19:39:22-07:00 America/Los_Angeles

Fix weird bug with TAP::Harness and print qq{ }

0.094     2015-07-12 16:28:33-07:00 America/Los_Angeles

Modify all POD so we can use Pod::Weaver

Retrograde code syntax so we can run under perls back to 5.6.0

0.092     2015-07-10 11:05:43-07:00 America/Los_Angeles

0.091     2015-07-10 10:54:22-07:00 America/Los_Angeles



lib/CMS/Drupal/Modules/MembershipEntity/Test.pm
package CMS::Drupal::Modules::MembershipEntity::Test;

# ABSTRACT: Exports some helper routines for testing

use strict;
use warnings;

use base "Exporter::Tiny";
our @EXPORT = qw/ build_test_db
                  build_and_validate_test_db
                  build_test_data /;

use Carp qw/ croak confess /;
use Test::More;
use Test::Group;
use File::Slurper qw/ read_text read_lines /;
use FindBin;
use DBD::SQLite;
use Time::Local;

sub build_and_validate_test_db {

  ## Reads the test data from .dat files and
  ## builds an in-memory SQLite database

  ## This package expects to be used in a program
  ## running in the test directory ...
  
  my $drupal = shift;
  my $dbh;

  subtest 'Built the test database' => sub {
    
    plan tests => 5;
    $dbh = $drupal->dbh( database => ':memory:',
                         driver   => 'SQLite' );

    isa_ok( $dbh, 'DBI::db', 'Got a valid $dbh' );

    subtest 'Created the test database tables.' => sub {
      plan tests => 4;
      for (split( /\n{2,}/, read_text("$FindBin::Bin/data/test_db.sql") )) {
        my $rv = $dbh->do($_);
        isnt( $rv, undef, 'Added a table to the test database' );
      }   
    };  

    #########
    # Populate the test database
  
    # First we have to have a default type
    my $add_type = qq/
      INSERT INTO membership_entity_type (type, label, weight, description, data, status, module)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    /;
  
    my @fields = split(',', read_text("$FindBin::Bin/data/test_types.dat")) or croak; # This file must have only ONE line
  
    my $add_type_rv = $dbh->do( $add_type, {}, @fields, undef );
    cmp_ok( $add_type_rv, '>', 0, 'Populate the membership_entity_type table with a default type' );
  
    ## Now add Memberships from the data file
    my $add_mem = qq/
      INSERT INTO membership_entity (mid, member_id, type, uid, status, created, changed)
      VALUES ( ?, ?, ?, ?, ?, ?, ?)
      /;
  
    test 'Populate the membership_entity table with test data' => sub {
      for ( read_lines("$FindBin::Bin/data/test_memberships.dat" ) ) {
        my @fields = split(',');
        my $add_mem_rv = $dbh->do( $add_mem, {}, @fields );
        cmp_ok( $add_mem_rv, '>', 0, "Added a Membership for mid $fields[0]" );
      }
    };
  
    ## Now add Membership Terms from the data file
    my $add_term = qq/
      INSERT INTO membership_entity_term(id, mid, status, term, modifiers, start, end )
      VALUES (?, ?, ?, ?, ?, ?, ?)
      /;
  
    test 'Populate the membership_entity_term table with test data' => sub {
      for ( read_lines("$FindBin::Bin/data/test_terms.dat" ) ) {
        my @fields = split(',');
        my $add_term_rv = $dbh->do( $add_term, {}, @fields );
        cmp_ok( $add_term_rv, '>', 0, "Added a Term for $fields[0]" );
      }
    };
  }; # done building the test DB

  $dbh->sqlite_backup_to_file("$FindBin::Bin/data/.DB.sqlite");

  return $dbh;
}

sub build_test_db {

  ## No testing here!

  ## Reads the test data from .dat files and
  ## builds an in-memory SQLite database


  my $drupal = shift;
  my $dbh = $drupal->dbh( database => ':memory:',
                          driver   => 'SQLite' );

  for (split( /\n{2,}/, read_text("$FindBin::Bin/data/test_db.sql") )) {
    my $rv = $dbh->do($_);
  }

  # First we have to have a default type
  my $add_type = qq/
    INSERT INTO membership_entity_type (type, label, weight, description, data, status, module)
    VALUES (?, ?, ?, ?, ?, ?, ?)
  /;

  my @fields = split(',', read_text("$FindBin::Bin/data/test_types.dat")) or croak; # This file must have only ONE line

  my $add_type_rv = $dbh->do( $add_type, {}, @fields, undef );

  ## Now add Memberships from the data file
  my $add_mem = qq/
    INSERT INTO membership_entity (mid, member_id, type, uid, status, created, changed)
    VALUES ( ?, ?, ?, ?, ?, ?, ?)
    /;

  for ( read_lines("$FindBin::Bin/data/test_memberships.dat" ) ) {
    my @fields = split(',');
    my $add_mem_rv = $dbh->do( $add_mem, {}, @fields );
  }

  ## Now add Membership Terms from the data file
  my $add_term = qq/
    INSERT INTO membership_entity_term(id, mid, status, term, modifiers, start, end )
    VALUES (?, ?, ?, ?, ?, ?, ?)
    /;

  for ( read_lines("$FindBin::Bin/data/test_terms.dat" ) ) {
    my @fields = split(',');
    my $add_term_rv = $dbh->do( $add_term, {}, @fields );
  }

  $dbh->sqlite_backup_to_file( "$FindBin::Bin/data/.DB.sqlite" );

  return $dbh;
} 

############################

sub build_test_data {

  my @mids = @_;
  my %include;
  for( @mids ) {
    $include{ $_ }++;
  }
 
  my %membs;
  my %terms;

  for ( read_lines("$FindBin::Bin/data/test_memberships.dat" ) ) {
    my @fields = split(',');
    if (scalar @mids > 0) { next unless exists $include{ $fields[0] }; }
    $membs{ $fields[0] } = { 'mid'       => $fields[0],
                             'member_id' => $fields[1],
                             'type'      => $fields[2],
                             'uid'       => $fields[3],
                             'status'    => $fields[4],
                             'created'   => $fields[5],
                             'changed'   => $fields[6] };
  }

  my %term_count;

  for ( read_lines("$FindBin::Bin/data/test_terms.dat" ) ) {
    my @fields = split(',');
    if (scalar @mids > 0) { next unless exists $include{ $fields[1] } };
    $term_count{ $fields[1] }++;
    for (5..6) {
      my @datetime = reverse (split /[-| |:]/, $fields[ $_ ]);
      $datetime[4]--;
      $fields[ $_ ] = timelocal( @datetime );
    }
    $terms{ $fields[0] } = bless(
                          { 'tid'            => $fields[0],
                            'mid'            => $fields[1],
                            'status'         => $fields[2],
                            'term'           => $fields[3],
                            'modifiers'      => $fields[4],
                            'start'          => $fields[5],
                            'end'            => $fields[6],
                            'array_position' => $term_count{ $fields[1] } },
                          'CMS::Drupal::Modules::MembershipEntity::Term' );
  }

  while ( my ($tid, $term) = each %terms ) {
    $membs{ $term->{'mid'} }->{'terms'}->{ $tid } = $term;
  }

  foreach my $mem ( keys %membs ) {
    $membs{ $mem } = bless( $membs{ $mem }, 'CMS::Drupal::Modules::MembershipEntity::Membership' );
  }

  return (scalar keys %membs == 1) ? @membs{ keys %membs } : \%membs;
}

1; # End package
__END__

=head1 SYNOPSIS

  use Test::More;
  use CMS::Drupal;
  use CMS::Drupal::Modules::MembershipEntity::Test qw/ build_test_db build_test_data/;

  my $drupal = CMS::Drupal->new;

  my $dbh = build_test_db( $drupal );

  my $ME = CMS::Drupal::Modules::MembershipEntity->new( dbh => $dbh );

  my $hashref = $ME->fetch_memberships;
  my $cmp_data = build_test_data;

  # or:
 
  my $hashref = $ME->fetch_memberships('1234', '5678');
  my $cmp_data = build_test_data('1234', '5678');

  is_deeply($hashref, $cmp_data, 'Data matches'); 

=head1 DESCRIPTION

Use this module when testing the CMS::Drupal::Modules::MembershipEntity modules.

=method build_and_validate_test_db

This method returns a database handle ($dbh) that is connected to an in-memory SQLite database.
The database is built by this method using data files that must be contained in the same directory that
the calling script lives in.

The method takes one argument, which must be a $drupal object. This is because it calls $drupal->dbh()
to generate its database handle, which, because we are using SQLite, contains the database inside the 
handle itself.

The files are:

=for :list
* B<test_db.sql>
* B<test_types.dat>
* B<test_memberships.dat>
* B<test_terms.dat>

Note that this method uses Test::More and Test::Group itself to report success/failures in building
the test database. So in your scipt that calls this method you should add one additional test to
your plan.

=method build_test_db

This method does the same as the previous one except it does not run tests itself, in other words
it builds the test db and the test data silently.

=method build_test_data

This method returns a data structure containing the Memberships as they would be returned by
CMS::Drupal::Modules::MembershipEntity::fetch_memberships(). It creates the data structure
by parsing the same files that were used to build in test database.

The method takes an optional single argument, which is an array containing a list of B<mid>s.
Only the Memberships associated with the B<mid>s provided will be included in the data
returned.

  $cmp_data = build_test_data('1234', '5678');

The data structure is a hashref of hashrefs (Membership objects, indexed by mid, containing
among their attributes an array of hashrefs (Membership Term objects) ...

  '4086' => bless( {
                    'mid'       => '4086',
                    'uid'       => '12305',
                    'status'    => '1',
                    'member_id' => '01252',
                    'terms' => {
                                '4088' => bless( {
                                                  'mid'            => '4086',
                                                  'array_position' => 2,
                                                  'status'         => '1',
                                                  'modifiers'      => 'a:0:{}',
                                                  'end'            => 1448611200,
                                                  'start'          => 1354089600,
                                                  'term'           => 'import',
                                                  'tid'            => '4088'
                                                 }, 'CMS::Drupal::Modules::MembershipEntity::Term' ),
                                '3920' => bless( {
                                                  'mid'            => '4086',
                                                  'array_position' => 1,
                                                  'status'         => '0',
                                                  'modifiers'      => 'a:0:{}',
                                                  'end'            => 1403247600,
                                                  'start'          => 1308639600,
                                                  'term'           => 'import',
                                                  'tid'            => '3920'
                                                 }, 'CMS::Drupal::Modules::MembershipEntity::Term' )
                               },
                    'type' => 'membership',
                    'created'   => '1354086000',
                    'changed'   => '1400604379'
                   }, 'CMS::Drupal::Modules::MembershipEntity::Membership' ),
 
=head1 SEE ALSO

=for :list
* L<CMS::Drupal::Modules::MembershipEntity::Membership|CMS::Drupal::Modules::MembershipEntity::Membership>
* L<CMS::Drupal::Modules::MembershipEntity::Term|CMS::Drupal::Modules::MembershipEntity::Term>
* L<CMS::Drupal::Modules::MembershipEntity|CMS::Drupal::Modules::MembershipEntity>
* L<CMS::Drupal|CMS::Drupal>



lib/CMS/Drupal/Modules/MembershipEntity.pm
package CMS::Drupal::Modules::MembershipEntity;

# ABSTRACT: Perl interface to Drupal MembershipEntity entities

use strict;
use warnings;

use Moo;
use Types::Standard qw/ :all /;
use Time::Local;
use Carp qw/ carp croak /;

use CMS::Drupal::Modules::MembershipEntity::Membership;
use CMS::Drupal::Modules::MembershipEntity::Term;

has dbh    => ( is => 'ro', isa => InstanceOf['DBI::db'], required => 1 );
has prefix => ( is => 'ro', isa => Maybe[StrMatch[ qr/ \w+_ /x ]] );

sub fetch_memberships {

  my $self = shift;

  ## We accept a list of mids as an optional parameter
  my @mids = @_;

  my $WHERE = ' ';
 
  if ( @mids ) {
    if ( scalar @mids < 1 ) {
      carp 'Empty array passed to fetch_memberships() ... returning all Memberships';
    }

    if ( ! grep { /all/ } @mids ) {
       # ^ in that case no WHERE clause

      for ( @mids ) {
        # Let's be real strict about what we try to pass in to the DBMS
        croak 'FATAL: Invalid mid (must be all ASCII digits).'
          unless /^[0-9]+$/;
      
        $WHERE = 'WHERE ';
        $WHERE .= "mid = '$_' OR " for @mids;
        $WHERE =~ s/ OR $//;
      }
    }
  }

  my $prefix = ( $self->{'prefix'} || '' );

  my %temp;
  my %memberships;

  ## Get the Membership info
  my $sql = qq|
    SELECT mid, member_id, type, uid, status, created, changed
    FROM ${prefix}membership_entity
    $WHERE
  |;
  
  my $sth = $self->{'dbh'}->prepare( $sql );
  $sth->execute;
  
  my $results = $sth->fetchall_hashref('mid');
  foreach my $mid (keys( %{ $results } )) {
    $temp{ $mid } = $results->{ $mid };
  }
  
  ## Get the Membership Term info
  #  Use the $WHERE clause from the optional mids parameter
  my $sql2 = qq|
    SELECT id as tid, mid, status, term, modifiers, start, end
    FROM ${prefix}membership_entity_term
    $WHERE
    ORDER BY start
  |;
  
  my $sth2 = $self->{'dbh'}->prepare( $sql2 );
  $sth2->execute;

  my %term_count; # used to track array position of Terms

  while( my $row = $sth2->fetchrow_hashref ) {
    
    ## Shouldn't be, but is, possible to have a Term with no
    ## start or end date
    if ( not defined $row->{'start'} or not defined $row->{'end'} ) {
      carp "MISSING DATE: tid[ $row->{'tid'} ] " .
           "(uid[ $temp{ $row->{'mid'} }->{'uid'} ]) has no start " .
           "or end date defined. Skipping ...";
      next;
    }

    ## Shouldn't be, but is, possible to have a Term with no
    ## corresponding Memberships
    if ( not defined $temp{ $row->{'mid'} } ) {
      carp "TERM WITH NO MEMBERSHIP: tid[ $row->{'tid'} ] " .
           "has no corresponding Membership. Skipping ...";
      next;
    }

    ## convert the start and end to unixtime
    for (qw/ start end /) {
      my @datetime = reverse ( split /[-| |:]/, $row->{ $_ } );
      $datetime[4]--;
      $row->{ $_ } = timelocal( @datetime );
    } 

    ## Track which of the Membership's Terms this is
    $term_count{ $row->{'mid'} }++;
    $row->{'array_position'} = $term_count{ $row->{'mid'} };
    
    ## Instantiate a MembershipEntity::Term object for each
    ## Term now that we have the data
    my $term = CMS::Drupal::Modules::MembershipEntity::Term->new( $row );
    $temp{ $row->{'mid'} }->{'terms'}->{ $row->{'tid'} } = $term;
  }

  ## Instantiate a MembershipEntity::Membership object for each
  ## Membership now that we have the data
  foreach my $mid(keys( %temp )) {
    
    ## Shouldn't be, but is, possible to have a Membership with no Term
    if (not defined $temp{ $mid }->{'terms'}) {
      carp "MISSING TERM: mid[ $mid ] (uid[ $temp{ $mid }->{'uid'} ]) " .
           "has no Membership Terms. Skipping ...";
      next;
    }

    $memberships{ $mid } =
    CMS::Drupal::Modules::MembershipEntity::Membership->new( $temp{ $mid } );
  }
  
  $self->{'_memberships'} = \%memberships;

  return (scalar keys %memberships == 1) ?
       @memberships{ keys %memberships } :
                            \%memberships;
}

1; ## return true to end package CMS::Drupal::Modules::MembershipEntity
__END__

=head1 SYNOPSIS

  use CMS::Drupal::Modules::MembershipEntity;

  my $ME = CMS::Drupal::Modules::MembershipEntity->new( { dbh => $dbh } );

  my $href = $ME->fetch_memberships( 'all' );

  # or:
  my $href = $ME->fetch_memberships( @list );

  foreach my $mid ( sort keys %{ $hashref } ) {
    my $mem = $hashref->{ $mid };
   
    print $mem->type;
    send_newsletter( $mem->uid ) if $mem->active;
   
    # etc ...
  }

Or, for a single Membership:

   my $mem = $ME->fetch_memberships( 123 );

   print $mem->type;
   send_newsletter( $mem->uid ) if $mem->active;
        
   # etc ...

=head1 USAGE

This package provides easy access to Perl objects representing Membership
Entity memberships and their terms. Rather than creating those objects
directly, you should allow this module to do so.

For each Membership that you want, you can fetch a Membership object, which
contains at least one Term object, so you have access to all the methods
you can use on your Membership and its Terms.

For this reason the methods actually provided by the submodules are documented
here.

=method fetch_memberships

This method returns either a hashref containing Membership objects indexed by
B<mid>, or a single Membership object (if it was called with a single B<mid>).

When called with the argument 'all', the hashref contains all Memberships in 
the Drupal database, which might be too much for your memory if you have 
lots of them.

When called with an array containing B<mid>s, the hashref will contain an 
object for each mid in the array.

When called with a single B<mid>, the method will return a single object
(no hashref).

  # Fetch a single Membership
  my $mem = $ME->fetch_memberships( 1234 ); 

  # Fetch a set of Memberships
  my $hashref = $ME->fetch_memberships( 1234, 5678 );

  # Fetch a set of Memberships using a list you prepared elsewhere
  my $hashref = $ME->fetch_memberships( @list );

  # Fetch all your Memberships
  my $hashref = $ME->fetch_memberships('all');

  # Same thing but with a warning
  my $hashref = $ME->fetch_memberships();

IMPORTANT: If you have bad records in your Drupal database, the module will
print a warning and skip the record. This happens when there are no Terms
belonging to the Membership, or when the Term is missing a start date or end
date. You should immediately normalize your data! This issue will also
cause installation testing to fail if you have configured your environment
to test against your real Drupal database.

=head2 Memberships

This module uses CMS::Drupal::Modules::MembershipEntity::Membership so you
don't have to. The methods shown below are actually in the latter 
module where they are documented completely.

=head3 Attributes

You can directly access all the Membership's attributes as follows:

  $mem->attr_name

Where attr_name is one of:

  mid           
  member_id
  type
  uid
  status
  created
  changed

There is also another attribute 'terms', which contains a hashref of Term
objects, indexed by B<tid>. Each Term can be accessed by the methods described
in the Membership Terms section below.

=head3 Membership methods

Once you have the Membership object, you can call some methods on it:

  print 'User ' . $mem->uid . ' is in good standing' if $mem->is_active;
  print $mem->mid . ' has already renewed' if $mem->has_renewal;

Methods are:

=for :list
* is_active()
* is_expired()
* is_cancelled()
* is_pending()
* has_renewal()
* current_was_renewal()

=head2 Membership Terms

This module uses CMS::Drupal::Modules::MembershipEntity::Term so you
don't have to. The methods described below are actually in the latter
module where they are documented compeletely.

  while ( my ($tid, $term) = each %{ $mem->{'terms'} } ) {
    # do something ...
  }

=head3 Attributes

You can directly access all the Term's attributes as follows:

  $term->attr_name;

Where attr_name is one of:

  tid
  mid
  status
  term
  modifiers
  start
  end

There is also another attribute, 'array_position', which is used to determine if
the Term is a renewal, etc.

=head3 Membership Term methods

  print 'This is a live one' if $term->is_current;
  push @renewals, $term->tid if $term->was_renewal;

Methods are:

=for :list
* is_active()
* is_current()
* is_future()
* was_renewal()

=head1 SEE ALSO

=for :list
* L<CMS::Drupal::Modules::MembershipEntity::Membership|CMS::Drupal::Modules::MembershipEntity::Membership>
* L<CMS::Drupal::Modules::MembershipEntity::Term|CMS::Drupal::Modules::MembershipEntity::Term>
* L<CMS::Drupal|CMS::Drupal>



lib/CMS/Drupal/Modules/MembershipEntity/Cookbook.pm
package CMS::Drupal::Modules::MembershipEntity::Cookbook;

# ABSTRACT: Guide and tutorials for using the Perl-Drupal Membership Entity interface

use strict;
use warnings;

=head1 SYNOPSIS

This manual contains a collection of tutorials and tips for using
the CMS::Drupal::Modules::MembershipEntity distribution.

=head1 DESCRIPTION

The individual packages in the CMS::Drupal::Modules::MembershipEntity
distribution each have their own POD of course, but the author hopes
that this documentation will help a new user put it all together.
Maybe you are a non-programmer or a non-Perl user and you are here
because you use Drupal's MembershipEntity modules and you need the
additional tools this distribution provides.

=head2 Code examples

In the interests of brevity and readability I have omitted the standard
opening lines from the code samples below. If you are copy-pasting the
examples and trying them out on your system, you should prepend the
following to each snippet:

  #!/usr/bin/perl -w
  use strict;
  use feature 'say';

The examples also skip the "use Foo::Bar"" lines, except for those
examples in each section that specifically describe how to "use"
them. You'll need to include those lines in your code too!

Note that these examples use the feature "say" which became available in
Perl v5.10 ... you can of course replace with "print" if you like: I
prefer "say" in examples (and in my code!) because you can omit the
newlines and their quotation marks.

=head1 INSTALLATION AND TESTING

Install the modules however you normally do. The easiest way is to get
them from CPAN:

  $ cpan install CMS::Drupal
  $ cpan install CMS::Drupal::Modules::MembershipEntity

If you want the modules to test themselves against your Drupal database
you will need to set the DRUPAL_TEST_CREDS environment variable as 
described in the Testing section of the documentation for the parent
CMS::Drupal module. Essentially you will need to provide at least
the database name and driver, so your minimum testing config would be:

  $ set DRUPAL_TEST_CREDS=database,foo,driver,Pg

or something similar.

If you come to this tutorial after you've installed the modules you
can still test by finding the CPAN build directory and just running

  $ perl t/02_valid_drupal.t

=head1 USING THE MODULES

You'll need to load the libraries in any script or program you write.
First you need to get a connection to your Drupal database. This is
done through the L<CMS::Drupal|CMS::Drupal> parent module:

  #!/usr/bin/perl -w
  use strict;
  use CMS::Drupal;
 
  my $drupal = CMS::Drupal->new;
  my $dbh    = $drupal->dbh(
    'database' => "my_db",
    'driver'   => "mysql",
    'username' => "my_user",
    'password' => "my_password"
  );
  
Once you have a DB connection you can make use of the Membership Entity
modules. Again, you'll have to load them with use():

  #!/usr/bin/perl -w
  use strict;
  use CMS::Drupal;
  use CMS::Drupal::Modules::MembershipEntity;

  my $drupal = CMS::Drupal->new;
  my $dbh    = $drupal->dbh(
    'database' => "my_db",
    'driver'   => "mysql",
    'username' => "my_user",
    'password' => "my_password"
  );
  
  my $ME = CMS::Drupal::Modules::MembershipEntity->new( dbh => $dbh );
  my $h  = $ME->fetch_memberships('all');

Now you have a hashref containing all our Memberships, and you can work
with each one:

  foreach my $mid ( keys %{ $h } ) {
    
    if ( $h->{ $mid }->is_active ) {
      $active_count++;
    }
    
    # pass the Membership object to a sub you wrote
    frobinate_member( $h->{ $mid } ); 

  }

If you just want to fetch one Membership, just pass a single mid to the 
fetch_memberships() method, and you'll get a single object returned 
instead of a hashref:

  my $mem = $ME->fetch_memberships( 12345 );
  print $mem->type;
  frobinate_member( $mem );

=head1 DATA ANALYSIS

The creation of this distribution was originally motivated by the lack of
tools for doing even rudimentary data analysis in Drupal's MembershipEntity
modules. This section explains the various ways it can help you with that.

=head1 Stats.pm

The main tool for analyzing your Membership base is the module
L<CMS::Drupal::Modules::MembershipEntity::Stats|Stats>.

=head2 Usage and Importing Methods

Start out by using Stats.pm in your program after you have loaded
the MembershipEntity modules as decribed above.

Notice that because Stats.pm exports all its useful methods, you can
import them into MembershipEntity.pm and thus make them available in
your $ME object:

  #!/usr/bin/perl -w
  use strict;
  use CMS::Drupal;
  use CMS::Drupal::Modules::MembershipEntity;
  use CMS::Drupal::Modules::MembershipEntity::Stats
       { into => 'CMS::Drupal::Modules::MembershipEntity' };
  
  my $drupal = CMS::Drupal->new;
  my $dbh    = $drupal->dbh;
  my $ME = CMS::Drupal::Modules::MembershipEntity->new( dbh => $dbh );
  $ME->fetch_memberships('all');

  my $active = $ME->count_active_memberships;
  say "We have $active active Memberships";

=head2 Types of methods

The methods in Stats.pm can be grouped into three basic categories,
which are signalled through the prefix of the methods' names:

=for :list
* count_
These methods may be called with or without a set of Memberships. If
called without a set (see below), they will operate on all your
Memberships.
* count_set_
These methods B<must> be called with a set of Memberships.
* count_daily_
These methods B<must> be called with a range of dates (see below), and
may be called with a set of Memberships (see below).

=head2 Working with a set of Memberships

Methods whose names begine with the prefix 'count_set_' must be called
with a set of Memberships, even if you want all Memberships included.
The set of Memberships is already contained in your MembershipEntity
object if you called fetch_memberships() on your MembershipEntity object
as described above:

  my @mids = ( 123, 456, 665, 667 );
  $ME->fetch_memberships( @mids );

or

  $ME->fetch_memberships('all');

If you do not call fetch_memberships() first, calling a method whose
name begins with the prefix 'count_set_' will cause your program to
die.

On the other hand, if you want to limit the statistics returned by
a 'count_' method or 'count_daily_' method to a subset of your 
Memberships, you can do so by first calling fetch_memberships() with
a list of mids.

=head2 Performance

For reasons of performance, and especially if you have a lot of 
Memberships, you should use Stats.pm B<without> first calling
fetch_memberships(), if you do not need to use any of the 
count_set_*() methods. This is because it will get its counts
by querying the database directly rather than by instantiating several
objects for each Membership.

If you are working with stats for your entire Membership base, you
should therefore not call fetch_memberships(). If you want to use
the count_set_*() methods on all your Memberships, you will have to 
call fetch_memberships('all') first, but you can then remove the set
and go back to the fast methods as shown below

  my $all = $ME->count_all_memberships; # fast
  
  # load the Memberships into a hashref
  $ME->fetch_memberships('all');
 
  # get a count that calls methods on the objects
  my $num = $ME->count_set_were_renewal_memberships();
  
  delete $ME->{'_hashref'};

  my $active = $ME->count_active_memberships; # back to the fast way

=head2 Working with dates

The methods in Stats.pm whose names begin with the prefix count_daily_
return counts for a date or range of dates. They can be optionally
limited to search on a set of Memberships if the fetch_memberships()
method is called in advance. These methods take dates in ISOish
format, i.e. something like:

  2001-01-01T12:00:00

(called ISO-ish because there is no time zone, as the ISO 8601 format
specifies).

You can use these methods to look at a date in the past:

  my $num =
    $ME->count_daily_expired_memberships('2015-06-15T00:00:00');

or at a range of dates:

  my $counts =
    $ME->count_daily_expired_memberships('2015-06-15T00:00:00',
                                         '2015-07-15T00:00:00',
                                         '2015-08-15T00:00:00');
  
  while (my ($date, $count) = each %{ $counts }) {
    $quarterly_total += $count;
    ...
  }

=head3 What time to use?

If you are forensically building a record of your Memberships, you
should probably set the time element of your date(s) to 00:00:00,
since the searchlooks for Terms with a start date <= and an end
date > the date given, and this strategy will give a daily report
that is closest to the historical truth if you store your statistics
indexed by a date.

=head3 Building a range of dates

Working with dates can be cumbersome, and typing them manually is
prone to errors. The Stats.pm module provides a method to build an
array of datetime strings that you can pass to its other methods.

You call it with either one or two dates: the first is required
and is the start date in the range. The second argument, if provided,
is used for the end date in the range. If no end date is provided,
the module will use today's date. Times are set to 00:00:00.

  my @dates = $ME->build_date_range('2014-01-01','2014-12-31');
  # one year's worth of dates

  my @dates = $ME->build_date_range('2001-01-01);
  # Every date in the millenium so far

=head3 An example

Here's an example of a simple program that reports the previous
week's statistics.

  #!/usr/bin/perl -w
  use strict;
  use DateTime;
  use CMS::Drupal;
  use CMS::Drupal::Modules::MembershipEntity;
  use CMS::Drupal::Modules::MembershipEntity::Stats
        { into => 'CMS::Drupal::Modules::MembershipEntity' };

  my $drupal = CMS::Drupal->new;
  my $dbh    = $drupal->dbh;
  my $ME = CMS::Drupal::Modules::MembershipEntity->new({dbh => $dbh});

  my $today      = DateTime->now()->set_time_zone('UTC');
  my $a_week_ago = $today->subtract( days => 6 )->ymd();
                   # 6 days because the set includes today

  my @days = @{ $ME->build_date_range( $a_week_ago ) };

  my $report = <<EOT;
  MembershipEntity Report
  ---------------------------
  Date        Exp New Active
  ---------------------------
  EOT

  my $exp = $ME->count_daily_term_expirations( @days );
  my $new = $ME->count_daily_new_terms( @days );
  my $act = $ME->count_daily_active_memberships( @days );

  foreach my $date ( @days ) { 
    my @line = substr $date, 0, 10;  
    for ( $exp, $new, $act ) { 
      push @line, $_->{ $date };
    }
    $report .= (join ' | ', @line) . "\n";
  }
  $report .= '-' x 25 . "\n";
  
  print $report;

  __END__

This program outputs something like:

  MembershipEntity Report
  ---------------------------
  Date        Exp New Active
  ---------------------------
  2015-07-09 | 0 | 0 | 580
  2015-07-10 | 0 | 0 | 580
  2015-07-11 | 0 | 1 | 580
  2015-07-12 | 0 | 1 | 580
  2015-07-13 | 1 | 0 | 581
  2015-07-14 | 0 | 1 | 580
  2015-07-15 | 1 | 0 | 580
  2015-07-16 | 0 | 0 | 580
  -------------------------
 

Now all you have to do is get the data for multiple weeks, sum
up the daily totals for the week, and you have the beginnings
of a useful report!

=head3 Automated daily report

If you want to build up a statistical record about your Memberships,
you'll probably want to export counts from the database into a
spreadsheet or other local database, so you can analyse them and make
graphs and so on.

The simplest way to do this is to write a program similar to the one
above and run it from cron every day. The Stats.pm module provides
a handy method for retrieving yesterday's data.

If you call $ME->report_yesterday() you'll get an anonymous hash
like this:

 {
  'count_daily_term_expirations' => '1',
  'count_daily_new_memberships' => '2',
  'count_daily_were_renewal_memberships' => '224',
  'count_daily_total_memberships' => '1498',
  'count_daily_new_terms' => '3',
  'count_daily_term_activations' => '2',
  'count_daily_active_memberships' => '580',
  'count_daily_renewals' => '2',
  'date' => '2013-02-14T00:00:00'
 }

You can optionally pass an anonymous array of method names that you
want exlcuded from the output. Check the documentation for Stats.pm
for a list of all the count_daily_*() methods.

So now you can do something in your nightly program something like:

  ## Assumes you have a database with a table called 'daily'

  my $sql = qq/ 
    INSERT INTO daily
    ('date', 'active', 'expirations', 'new_terms', 'new_memberships', 'renewals')
    VALUES ( ?, ?, ?, ?, ?, ? ) 
  /;

  my $sth = $dbh->prepare( $sql );

  my @exclude = qw/ count_daily_total_memberships
                    count_daily_term_activations
                    count_daily_were_renewal_memberships /;

  my $yesterday = $ME->report_yesterday( \@exclude );

  $sth->execute( $yesterday->{'date'},
                 $yesterday->{'count_daily_active_memberships'},
                 $yesterday->{'count_daily_term_expirations'},
                 $yesterday->{'count_daily_new_terms'},
                 $yesterday->{'count_daily_new_memberships'},
                 $yesterday->{'count_daily_renewals'} );


. . . or of course you could send an email, update a text log, whatever...



=cut




1; # return true
__END__


